runtime/Cpp/runtime/src/atn/PredictionContext.cpp

```diff
@@ -252,130 +221,100 @@ Ref<PredictionContext> PredictionContext::mergeArrays(const Ref<ArrayPredictionC
+          ++i;
+          ++j;
+      } else if (i->returnState < j->returnState) {
+          contexts.emplace_back(i->parent, i->returnState);
+          ++i;
+      } else {
+          contexts.emplace_back(j->parent, j->returnState);
+          ++j;
       }
-      i++; // hop over left one as usual
-      j++; // but also skip one in right side since we merge
-    } else if (a->returnStates[i] < b->returnStates[j]) { // copy a[i] to M
-      mergedParents[k] = a_parent;
-      mergedReturnStates[k] = a->returnStates[i];
-      i++;
-    }
-    else { // b > a, copy b[j] to M
-      mergedParents[k] = b_parent;
-      mergedReturnStates[k] = b->returnStates[j];
-      j++;
-    }
-    k++;
   }
-
-  // copy over any payloads remaining in either array
-  if (i < a->returnStates.size()) {
-    for (std::vector<int>::size_type p = i; p < a->returnStates.size(); p++) {
-      mergedParents[k] = a->parents[p];
-      mergedReturnStates[k] = a->returnStates[p];
-      k++;
-    }
-  } else {
-    for (std::vector<int>::size_type p = j; p < b->returnStates.size(); p++) {
-      mergedParents[k] = b->parents[p];
-      mergedReturnStates[k] = b->returnStates[p];
-      k++;
-    }
+  while (i != i_end) {
+      contexts.emplace_back(i->parent, i->returnState);
+      ++i;
   }
-
-  // trim merged if we combined a few that had same stack tops
-  if (k < mergedParents.size()) { // write index < last position; trim
-    if (k == 1) { // for just one merged element, return singleton top
-      Ref<PredictionContext> a_ = SingletonPredictionContext::create(mergedParents[0], mergedReturnStates[0]);
-      if (mergeCache != nullptr) {
-        mergeCache->put(a, b, a_);
-      }
-      return a_;
-    }
-    mergedParents.resize(k);
-    mergedReturnStates.resize(k);
+  while (j != j_end) {
+      contexts.emplace_back(j->parent, j->returnState);
+      ++j;
   }
 
-  Ref<ArrayPredictionContext> M = std::make_shared<ArrayPredictionContext>(mergedParents, mergedReturnStates);
-
   // if we created same array as a or b, return that instead
   // TO_DO: track whether this is possible above during merge sort for speed
-  if (*M == *a) {
-    if (mergeCache != nullptr) {
-      mergeCache->put(a, b, a);
-    }
-    return a;
-  }
-  if (*M == *b) {
-    if (mergeCache != nullptr) {
-      mergeCache->put(a, b, b);
-    }
-    return b;
+  if (antlrcpp::Arrays::equals(contexts, a->contexts)) {
+      if (mergeCache != nullptr) {
+          mergeCache->put(a, b, a);
+      }
+      return a;
   }
-
-  // ml: this part differs from Java code. We have to recreate the context as the parents array is copied on creation.
-  if (combineCommonParents(mergedParents)) {
-    mergedReturnStates.resize(mergedParents.size());
-    M = std::make_shared<ArrayPredictionContext>(mergedParents, mergedReturnStates);
+  if (antlrcpp::Arrays::equals(contexts, b->contexts)) {
+      if (mergeCache != nullptr) {
+          mergeCache->put(a, b, b);
+      }
+      return b;
```

Comment:
Here I am again at a point where I cannot understand what you changed and can only hope you know what you are doing. The patch creates bigger differences (also in naming) compared to the original Java code, but maybe this is the price we have to pay to make this more "C++-ish"...

Reply:
I renamed a few of the iterators and put back the original comments in this regard. Hopefully that makes it more clear what the lineage of the code is. In particular there is a new merge method specifically for merging a singleton into an array. The merge method for this is based on mergeArrays but by doing this there is no need to allocate a temporary ArrayPredictionContext in order to call mergeArrays.